kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/ksp/KspUtil.kt

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.moshi.kotlin.codegen.ksp
+
+import com.google.devtools.ksp.processing.KSPLogger
+import com.google.devtools.ksp.processing.Resolver
+import com.google.devtools.ksp.symbol.ClassKind
+import com.google.devtools.ksp.symbol.KSAnnotated
+import com.google.devtools.ksp.symbol.KSAnnotation
+import com.google.devtools.ksp.symbol.KSClassDeclaration
+import com.google.devtools.ksp.symbol.KSName
+import com.google.devtools.ksp.symbol.KSNode
+import com.google.devtools.ksp.symbol.KSType
+import com.google.devtools.ksp.symbol.KSTypeAlias
+import com.google.devtools.ksp.symbol.Origin.KOTLIN
+import com.google.devtools.ksp.symbol.Origin.KOTLIN_LIB
+import com.google.devtools.ksp.symbol.Visibility
+import com.google.devtools.ksp.symbol.Visibility.INTERNAL
+import com.google.devtools.ksp.symbol.Visibility.JAVA_PACKAGE
+import com.google.devtools.ksp.symbol.Visibility.LOCAL
+import com.google.devtools.ksp.symbol.Visibility.PRIVATE
+import com.google.devtools.ksp.symbol.Visibility.PROTECTED
+import com.google.devtools.ksp.symbol.Visibility.PUBLIC
+import com.squareup.kotlinpoet.AnnotationSpec
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.ksp.toClassName
+
+internal inline fun <reified T> Resolver.getClassDeclarationByName(): KSClassDeclaration {
+  return getClassDeclarationByName(T::class.qualifiedName!!)
+}
+
+internal fun Resolver.getClassDeclarationByName(fqcn: String): KSClassDeclaration {
+  return getClassDeclarationByName(getKSNameFromString(fqcn)) ?: error("Class '$fqcn' not found.")
+}
+
+internal fun KSClassDeclaration.asType() = asType(emptyList())
+
+internal fun KSClassDeclaration.isKotlinClass(resolver: Resolver): Boolean {
+  return origin == KOTLIN ||
+    origin == KOTLIN_LIB ||
+    hasAnnotation(resolver.getClassDeclarationByName<Metadata>().asType())
+}
+
+internal fun KSAnnotated.hasAnnotation(target: KSType): Boolean {
+  return findAnnotationWithType(target) != null
+}
+
+internal inline fun <reified T : Annotation> KSAnnotated.findAnnotationWithType(
+  resolver: Resolver,
+): KSAnnotation? {
+  return findAnnotationWithType(resolver.getClassDeclarationByName<T>().asType())
```

Comment:
KSP now has a `getAnnotationsByType(KClass)` method to obtain annotation instances. Don't know how it compares performance-wise to the way you are doing it, but may simplify things?

Reply:
Found a bug but fixed in the shading 5d1792a